### PR TITLE
[6983] Improve XB ease-of-use and resilience for JS apps programmatically registering elements for XB

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/6983

### CODE REVIEW

**Review by rart — CHANGES_REQUESTED** [review]
Is this finished? I don't see for example the recordIdByElementLookup being cleaned when deregister is called?

Also not sure I'm seeing work to make sure deregister would would even when the deferred request hasn't completed.

File: ui/guest/src/store/subjects.ts [hunk 1]
```diff
@@ -29,6 +29,8 @@ export const escape$ =
     ? new Subject<KeyboardEvent>()
     : fromEvent<KeyboardEvent>(document, 'keydown').pipe(filter((e) => e.key === 'Escape'));
 
+export const guestCheckIn$ = new ReplaySubject<boolean>(1);
```
Comment (jvega190) [inline ui/guest/src/store/subjects.ts:32]: There's one detail here (in the usage). We still need to listen for the `xb:loaded` so were sure we have `craftercms.xb`, then subscribe to `guestCheckIn$`

File: ui/guest/src/elementRegistry.ts [hunk 2]
```diff
@@ -154,7 +163,7 @@ export function register(payload: ElementRecordRegistration): number {
     // for the model to be loaded.
     if (isInheritedField(model.craftercms.id, fieldId)) {
       byPathFetchIfNotLoaded(model.craftercms.sourceMap?.[fieldId]).subscribe((response) => {
-        model$(response.craftercms.id)
+        deferredRegistrationSubscription = model$(response.craftercms.id)
```
Comment (rart) [inline ui/guest/src/elementRegistry.ts]: There are many calls to register done sequentially. A single subscription doesn't work. You need maybe a stream of deregistered ids to stop this one.

File: ui/guest/src/elementRegistry.ts [hunk 3]
```diff
@@ -47,6 +47,11 @@ let seq = 0;
 let db: LookupTable<ElementRecord> = {};
 // Lookup table of element record id arrays, indexed by iceId
 let registry: LookupTable<number[]> = {};
+// Lookup table of element record id, index by the element
+const recordIdByElementLookup = new Map<Element, number>();
+
+let registrationSubscription;
+let deferredRegistrationSubscription;
```
Comment (rart) [inline ui/guest/src/elementRegistry.ts]: What's the need for 2 subjects?

**Review by rart — CHANGES_REQUESTED** [review]
Let's add/move the `registerElements` and `deregisterElements` functions to the library (to the index)

File: ui/guest/src/elementRegistry.ts (hunk 3 above)
Comment (jvega190) [reply ui/guest/src/elementRegistry.ts]: There are two places where we have subscriptions, in the main register funcion (if main model is not fetched) and in the `completeRegistration` function (if is an inherited field, to load the model). What I did was to add a lookup for registrations (with the id as the key), and the content is an array of the subscriptions, considering that there may be up to 2 subscriptions. Then, when deregistering, delete all of the subscriptions for the id

File: ui/guest/src/index.tsx [hunk 4]
```diff
@@ -101,6 +103,36 @@ export function initExperienceBuilder(props: ExperienceBuilderProps) {
...
+        label: element.getAttribute('data-craftercms-label')
+      });
+    });
+  }
+}
+
+export function deregisterElements(container: HTMLElement) {
+  if (container) {
+    const articleElements = container.querySelectorAll('[data-craftercms-model-id]');
+    articleElements.forEach((element) => {
```
Comment (rart) [inline ui/guest/src/index.tsx]: These aren't articles here...
```suggestion
    const elements = container.querySelectorAll('[data-craftercms-model-id]');
    elements.forEach((element) => {
```

File: ui/guest/src/elementRegistry.ts [hunk 5]
```diff
@@ -47,6 +47,10 @@ let seq = 0;
 let db: LookupTable<ElementRecord> = {};
 // Lookup table of element record id arrays, indexed by iceId
 let registry: LookupTable<number[]> = {};
+// Lookup table of element record id, index by the element
+const recordIdByElementLookup = new Map<Element, number>();
+// Subscription registry for deferred registrations
+let registrationSubscriptions = {};
```
Comment (rart) [inline ui/guest/src/elementRegistry.ts]: Pls apply discussed changes

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/guest/src/elementRegistry.ts [hunk 6]
```diff
@@ -41,12 +41,17 @@ import { notNullOrUndefined, nou, nullOrUndefined } from '@craftercms/studio-ui/
...
 
 let seq = 0;
 // Element record registry
 let db: LookupTable<ElementRecord> = {};
 // Lookup table of element record id arrays, indexed by iceId
 let registry: LookupTable<number[]> = {};
+// Lookup table of element record id, index by the element
+const recordIdByElementLookup = new Map<Element, number>();
+// Stream of ids being deregistered, used to cancel pending model fetch operations.
+const deregister$ = new Subject<number | string>();
```
Comment (rart) [inline ui/guest/src/elementRegistry.ts]:
```suggestion
const deregister$ = new Subject<string>();
```

File: ui/guest/src/elementRegistry.ts [hunk 7]
```diff
@@ -132,6 +141,7 @@ export function register(payload: ElementRecordRegistration): number {
       : Array.isArray(fieldId)
         ? fieldId
         : fieldId.split(',').map((str) => str.trim());
+  const terminator$ = deregister$.pipe(filter((_id) => _id === id));
```
Comment (rart) [inline ui/guest/src/elementRegistry.ts]:
```suggestion
  const terminator$ = deregister$.pipe(filter((_id) => _id === String(id)));
```

**Review by rart — CHANGES_REQUESTED** [review]
Pls remember to port changes to 4.1 PR too

**Review by rart — APPROVED** [review]

**Pull request merged**